<table style="float:left; border:none">
   <tr style="border:none; background-color: #ffffff">
       <td style="border:none">
           <a href="https://praktikum.yandex.ru/">     
           <img 
               src="https://yastatic.net/q/praktikum/v0.116.12-fix3-1580741332/static/apple-touch-icon.png" 
               style="width:50px"
           >
           </a>    
       </td>
       <td style="border:none">
           <h1>Яндекс.Практикум</h1>
       </td>
   </tr>
</table>

<div style="float:right;"><h2>Выпускной проект<p>
    Проект по SQL</h2></div>

<a name="000"></a>

<div class="col-md-3">
  <ul class="nav flex-column nav-pills">
    <li class="active"><a class="nav-link" id="home-tab" data-toggle="pill" href="#describe" role="tab" aria-controls="describe" aria-selected="true">Описание проекта</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#contents" role="tab" aria-controls="contents" aria-selected="false">Оглавление</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#step1" role="tab" aria-controls="step1" aria-selected="false">Описание данных</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#output11" role="tab" aria-controls="output11" aria-selected="false">Посчитайте затраты на каждый рекламный источник.</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#output12" role="tab" aria-controls="output12" aria-selected="false">Посчитайте выручку от заказов, сделанных с настольных ПК (тип устройства desktop)</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#output13" role="tab" aria-controls="output13" aria-selected="false">Ввод новой метрики эффективности рекламы - дневные затраты на рекламу на пользователя по источнику</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#output111" role="tab" aria-controls="output111" aria-selected="false">Финальные выводы</a></li></ul>
      </div>


<div class="tab-content">
<div id="describe" class="tab-pane fade in active">
    <h3 style="margin-left:270px">Описание проекта</h3>
    <p style="margin-left:270px">
Вы аналитик в онлайн-маркетплейсе «Заберу». Это настоящий рай для любителей бесполезных сувениров — там можно приобрести товары ручной работы от небольших производителей. «Заберу» проводит рекламные акции длительностью 1 день. Руководитель отдела маркетинга хочет узнать, насколько эффективны такие кампании.<br>
<br></p>
<ul style="margin-left:270px">    
    <li>Посчитайте затраты на каждый рекламный источник.</li>
    <li>Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег. Сделайте выводы</li>
    <li>Введём новую метрику эффективности рекламы - дневные затраты на рекламу на пользователя по источнику. Она рассчитывается для каждого отдельного дня и равна объёму денежных средств, которые были израсходованы на то, чтобы пользователь перешёл по рекламе на сайт из данного рекламного источника. Иными словами, метрика для каждого дня равна отношению рекламных затрат по данному источнику к количеству посетителей, перешедших на сайт с данного источника. Рассчитайте для каждого дня введённую метрику. Считайте, что целевое действие было произведено тогда, когда сесия началась. Например, если посетитель провёл на сайте несколько минут с 23:50 14 апреля по 0:20 15 апреля, то его нужно включать в рассчёт за 14 апреля</li>
    </ul>
  </div>

<div id="contents" class="tab-pane fade">
    <h3 style="margin-left:270px">Оглавление</h3>
    <p style="margin-left:270px">
        <a href="#001">Подключение к базе данных</a><br>
        <a href="#002">Исследуем данные</a><br>
        <a href="#003">Посчитайте затраты на каждый рекламный источник.</a><br>
        <a href="#004">Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop)</a><br>
        <a href="#005">Введём новую метрику эффективности рекламы</a><br>
        <a href="#summary">Общий вывод и рекомендации по работе с клиентами</a><br><br>
        <b>В каком виде должно быть решено задание?</b>
            <ul style="margin-left:270px"><li>Опишите цели исследования
                <li>Исследуйте таблицы (выведите первые строки)
                <li>Сделайте по одному SQL-запросу для решения каждой из трёх представленных задач
                <li>Выведите результаты каждого запроса в тетрадке
                <li>Опишите выводы по каждой из решённых задач
                <li>Доступы к базе данных
                <li>Следуйте инструкции подключения к базе данных</ul>
    <p style="margin-left:270px"><b>Примечания</b></p>
<p style="margin-left:270px">Не забывайте про функции, они сильно облегчат жизнь при выполнении запросов<br>
Ответ должен быть получен с использованием только языка запросов SQL.<br> Pandas допускается только для хранения и вывода результатов запросов.
</p>
  </div>

  <div id="step1" class="tab-pane fade">
      <h3 style="margin-left:270px">Описание данных</h3>
      <p>
      <ul style="margin-left:270px"> Таблица <b>visits</b> (лог сервера с информацией о посещениях сайта):
          <li><code>uid</code> — уникальный идентификатор пользователя
          <li><code>device</code> — категория устройства пользователя
          <li><code>startts</code> — дата и время начала сессии
          <li><code>endts</code> — дата и время окончания сессии
          <li><code>sourceid</code> — идентификатор рекламного источника, из которого пришёл пользователь
</ul>
    <ul style="margin-left:270px"> Таблица <b>orders</b> (информация о заказах):
          <li><code>uid</code> — уникальный идентификатор пользователя, который сделал заказ
          <li><code>buyts</code> — дата и время заказа
          <li><code>revenue</code> — выручка Заберу с этого заказа
</ul>
   <ul style="margin-left:270px"> Таблица <b>advertisment_costs</b> (информация о затратах на маркетинг):
          <li><code>sourceid</code> — идентификатор рекламного источника
          <li><code>dt</code> — дата
          <li><code>costs</code> — затраты на этот рекламный источник в этот день
</ul></div>

<div id="output11" class="tab-pane fade">
    <h3 style="margin-left:270px">Посчитайте затраты на каждый рекламный источник.</h3>
    <p style="margin-left:270px"> Всего 10 рекламных источников. Три из них отсутствуют. Возможно, это связано с технической ошибкой или отсутствием переходов с данных рекламных размещений. <br><br> Самые большие суммарные затраты — 141 тыс., связаны с рекламным размещением №3. Далее идёт рекламный источник №4 — 61 тыс. Примерно в равной степени самые минимальные затраты приходятся на рекламные размещения №9 и 10 — 5,5 и 5.8 тыс. соответственно</p>
  </div>

<div id="output12" class="tab-pane fade">
    <h3 style="margin-left:270px">Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег. Сделайте выводы</h3>
    <p style="margin-left:270px">
В целом затраты на рекламный источник № 3 не окупаются с декстопной версии платформы. Разница затрат равна 35,263 тыс. Конверсия выручки от затрат составляет примерно 75%. Несмотря на то, что большая часть вложений приходится на рекламное размещение № 3, на данный момент оно не приносит прибыли.<br><br> Возможно, дополнительная выручка закрывающая часть затрат связанных с рекламным размещением № 3 приходит из мобильной версии платформы. Поэтому, данному рекламному размещению принадлежит большая часть вложений бюджета маркетинга.

</p>
  </div>

  <div id="output13" class="tab-pane fade">
      <h3 style="margin-left:270px">Введём новую метрику эффективности рекламны - дневные затраты на рекламу на пользователя по источнику.</h3>
      <p>
    <p style="margin-left:270px">Судя по таблице, второй и третий рекламный источник по мобильной и десктопной версии не окупаются. Затраты превышают выручку. Тоже самое касается рекламного размещения №5 по мобилке.<br> <br>Рекламные размещения №9 и №10 — окупаются. Видимо, лучше всего отрабатывает рекламный источник №10 по десктопной версии.<br><br>Метрика отображает отношение рекламных затрат по данному рекламному источнику к количеству посетителей, перешедших на сайт с данного источника.
            </p>
  </div>
  
  <div id="output111" class="tab-pane fade">
      <h3 style="margin-left:270px">Финальные выводы</h3>
      <p>
        <p style="margin-left:270px">  
          Мы проанализировали стратегию рекламной кампании маркетплейса «Заберу». Она выстраивается на рекламных акциях длительностью <b>1 день</b>. В целом рекламная кампания показала свою неэффективность.<br><br>
Самые большие суммарные затраты — <b>141 тыс.</b>, связаны с рекламным размещением <b>№3</b>. Далее идёт рекламный источник №4 — 61 тыс. Примерно в равной степени самые минимальные затраты приходятся на рекламные размещения <b>№9 и 10</b> — 5,5 и 5,8 тыс. соответственно Основная часть рекламных размещений не окупается. <br><br><b>Видимо, лучше всего отрабатывает рекламный источник №10 по десктопной версии.</b></p>
            </div>    

 </div>  </div>  </div>  </div>  </div> 
  
  






### Подключение к базе данных <a name="001"></a>

Нам для подключения запросов необходима библиотека `psycopg2`, её можно установить через `!pip` - потребуется установка `psycopg2-binary`.
<br>Затем импортируем все библиотеки необходимые для выполнения проекта и создадим коннецкцию к базе.

In [1]:
!pip install psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable


In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine, MetaData, Table, \
insert, update, select, delete, desc, func, case, cast, Column, Integer, String
import psycopg2

db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])
# коннекция к базе
engine = create_engine(connection_string, connect_args={'sslmode':'require'}, echo=False)

SQL-запросы мы будем выполнять с помощью пандас, и чтобы не выводить операцию постоянно, сохраним функцию.

In [3]:
def sqlquery(query):
    return pd.io.sql.read_sql(query, con = engine)

### Исследуем данные<a name="002"></a>

Проверим названия таблиц в базе, выгрузим названия.

In [4]:
engine.table_names()

['orders', 'visits', 'advertisment_costs']

SQLAlchemy может использоваться для автоматической загрузки таблиц из базы данных, используя то, что называется отражением. Отражение - это процесс чтения базы данных и построения метаданных на основе этой информации. Это противоположно созданию таблицы вручную и очень полезно для работы с существующими базами данных.

Для отражения необходимо сначала импортировать и инициализировать объект `MetaData`. Объекты метаданных содержат информацию о таблицах, хранящихся в базе данных. Во время отражения объект `MetaData` будет автоматически заполняться информацией об отраженной таблице, поэтому нам нужно только инициализировать ее перед отражением, вызвав `MetaData ()`.

In [5]:
metadata = MetaData()

In [6]:
repr(Table('orders', metadata, autoload= True, autoload_with= engine))

"Table('orders', MetaData(bind=None), Column('id', INTEGER(), table=<orders>, primary_key=True, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.elements.TextClause object at 0x7f82152dfbe0>, for_update=False)), Column('buyts', TIMESTAMP(), table=<orders>), Column('revenue', MONEY(), table=<orders>), Column('uid', VARCHAR(length=30), table=<orders>), schema=None)"

In [7]:
repr(Table('visits', metadata, autoload= True, autoload_with= engine))

"Table('visits', MetaData(bind=None), Column('id', INTEGER(), table=<visits>, primary_key=True, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.elements.TextClause object at 0x7f8214ebccc0>, for_update=False)), Column('uid', VARCHAR(length=30), table=<visits>), Column('device', VARCHAR(length=10), table=<visits>), Column('endts', TIMESTAMP(), table=<visits>), Column('sourceid', SMALLINT(), table=<visits>), Column('startts', TIMESTAMP(), table=<visits>), schema=None)"

In [8]:
repr(Table('advertisment_costs', metadata, autoload= True, autoload_with= engine))

"Table('advertisment_costs', MetaData(bind=None), Column('id', INTEGER(), table=<advertisment_costs>, primary_key=True, nullable=False, server_default=DefaultClause(<sqlalchemy.sql.elements.TextClause object at 0x7f8214e517f0>, for_update=False)), Column('sourceid', SMALLINT(), table=<advertisment_costs>), Column('dt', TIMESTAMP(), table=<advertisment_costs>), Column('costs', MONEY(), table=<advertisment_costs>), schema=None)"

Ниже также еще один вариант посмотреть типы данных.

In [9]:
from sqlalchemy import inspect
inspector = inspect(engine)

for table_name in inspector.get_table_names():
    for column in inspector.get_columns(table_name):
#         print("Column: %s" % column['type'])
        print(column['name'], column['type'])

id INTEGER
buyts TIMESTAMP WITHOUT TIME ZONE
revenue MONEY
uid VARCHAR(30)
id INTEGER
uid VARCHAR(30)
device VARCHAR(10)
endts TIMESTAMP WITHOUT TIME ZONE
sourceid SMALLINT
startts TIMESTAMP WITHOUT TIME ZONE
id INTEGER
sourceid SMALLINT
dt TIMESTAMP WITHOUT TIME ZONE
costs MONEY


Выгрузим первые пять строк каждой таблицы.

In [10]:

orders_head = '''
SELECT *
FROM 
    orders
LIMIT 5;    
'''

sqlquery(orders_head)

id               buyts revenue                   uid
0   1 2017-06-01 00:10:00  $17.00  10329302124590727494
1   2 2017-06-01 00:25:00   $0.55  11627257723692907447
2   3 2017-06-01 00:27:00   $0.37  17903680561304213844
3   4 2017-06-01 00:29:00   $0.55  16109239769442553005
4   5 2017-06-01 07:58:00   $0.37  14200605875248379450

In [11]:
visits_head = '''
SELECT *
FROM 
    visits
LIMIT 5;    
'''

sqlquery(visits_head)

id                   uid   device               endts  sourceid  \
0   0  16879256277535980062    touch 2017-12-20 17:38:00         4   
1   1    104060357244891740  desktop 2018-02-19 17:21:00         2   
2   2   7459035603376831527    touch 2017-07-01 01:54:00         5   
3   3  16174680259334210214  desktop 2018-05-20 11:23:00         9   
4   4   9969694820036681168  desktop 2017-12-27 14:06:00         3   

              startts  
0 2017-12-20 17:20:00  
1 2018-02-19 16:53:00  
2 2017-07-01 01:54:00  
3 2018-05-20 10:59:00  
4 2017-12-27 14:06:00

In [12]:
costs_head = '''
SELECT *
FROM advertisment_costs
LIMIT 5;    
'''

sqlquery(costs_head)

id  sourceid         dt   costs
0   1         1 2017-06-01  $75.20
1   2         1 2017-06-02  $62.25
2   3         1 2017-06-03  $36.53
3   4         1 2017-06-04  $55.00
4   5         1 2017-06-05  $57.08

### Выводы по разделу

Мы посмотрели типы данных, информацию о таблицах (проверили все ли таблицы у нас были указаны в водных данных), а посмотрели на шапки таблиц.

Таблицы `orders` и `visits` у нас связаны по столбцу `uid`, а `visits` и `advertisment_costs` мы можем связать по `sourceid`.

### Посчитайте затраты на каждый рекламный источник
<a name="003"></a>

Затраты у нас содержатся в таблице `advertisment_costs`, а также там мы можем заметить что у нас есть сведения по источникам.

Выполним запрос где мы выберем столбцы `sourceid` и посчитаем к нему сумму по `costs`, группировку осуществим по источникам - `sourceid`.

In [13]:
query_1 = '''
SELECT sourceid, sum(costs) as sum
FROM advertisment_costs
GROUP BY sourceid
ORDER BY sum DESC;
'''

sqlquery(query_1)

sourceid          sum
0         3  $141,321.63
1         4   $61,073.60
2         5   $51,757.10
3         2   $42,806.04
4         1   $20,833.27
5        10    $5,822.49
6         9    $5,517.49

<div style="border:solid blue 2px; padding: 20px"> <h1 style="color:blue; margin-bottom:20px">Комментарий наставника</h1>

Тут всё посчитано верно, молодец


Всего 10 рекламных источников. Три из них отсутствуют. Возможно, это связано с технической ошибкой или отсутствием переходов с данных рекламных размещений.

Больше всего компания тратит на 3, 4 и 5 источники, в сумме примерно 250 тыс долларов.
Меньше всего затраты на 9 источник, 5,5 тыс долларов.

Посмотрим, сколько же всего пришло клиентов по каждому из источников.

In [14]:
query_2 = '''
SELECT sourceid, count(uid) as users
FROM visits
GROUP BY sourceid
ORDER BY users DESC;
'''

sqlquery(query_2)

sourceid   users
0         4  101653
1         3   85448
2         5   66711
3         2   47473
4         1   33943
5         9   13257
6        10   10005
7         7      36
8         6       6

### Вывод
Самые большие суммарные затраты — 141 тыс., связаны с рекламным размещением **№3**. Далее идёт рекламный источник №4 — 61 тыс. Примерно в равной степени самые минимальные затраты приходятся на рекламные размещения №9 и 10 — 5,5 тыс. и 5.8 тыс соответственно.

На 7 и 6 источники у компании нет затрат, скорей всего это сарафанное радио или какие-то публикации в интернете, за которые компания не вкладывала деньги.

С 3 4 и 5 источников приходят максимальное количество изеров, поэтому такие вложения оправданы на рекламу.

---

### Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег.
<a name="004"></a>

In [15]:
query3 = '''
SELECT sum(revenue) AS sum,
       max(revenue) AS max_revenue,
       min(revenue) AS min_revenue,
       max(revenue) - min(revenue) AS scope
       
FROM orders o

WHERE 
    o.uid IN (
              SELECT DISTINCT v.uid
              
              FROM visits v
              
              WHERE
                   v.device = 'desktop' AND
                   v.sourceid = '3');
'''

sqlquery(query3)

sum max_revenue min_revenue      scope
0  $106,058.86   $2,633.28       $0.00  $2,633.28

<div style="border:solid blue 2px; padding: 20px"> <h1 style="color:blue; margin-bottom:20px">Комментарий наставника</h1>

Результаты запроса правильные, супер!

Суммарная выручка составляет 106,058 тыс., максимальная — 2,633 $. Минимальная выручка равна 0. Возможно, это заказы пользователей, которые приобрели товары бесплатно (акция).

### Вывод
В целом затраты на рекламный источник № 3 не окупаются с декстопной версии платформы. Разница затрат равна 35,263 тыс. Конверсия выручки от затрат составляет примерно 75%. Несмотря на то, что большая часть вложений приходится на рекламное размещение № 3, на данный момент оно не приносит прибыли.

Возможно, дополнительная выручка закрывающая часть затрат связанных с рекламным размещением № 3 приходит из мобильной версии платформы. Поэтому, данному рекламному размещению принадлежит большая часть вложений бюджета маркетинга.

### Введём новую метрику эффективности рекламы - дневные затраты на рекламу на пользователя по источнику. 
<a name="005"></a>
Она рассчитывается для каждого отдельного дня и равна объёму денежных средств, которые были израсходованы на то, чтобы пользователь перешёл по рекламе на сайт из данного рекламного источника. Иными словами, метрика для каждого дня равна отношению рекламных затрат по данному источнику к количеству посетителей, перешедших на сайт с данного источника. Рассчитайте для каждого дня введённую метрику. Считайте, что целевое действие было произведено тогда, когда сесия началась. Например, если посетитель провёл на сайте несколько минут с 23:50 14 апреля по 0:20 15 апреля, то его нужно включать в рассчёт за 14 апреля

In [16]:

query4 = ''' SELECT
                        new_visits.date_revenue AS date,
                        new_visits.device,
                        new_visits.sourceid,
                        SUM(new_visits.revenue) AS total_revenue,
                        SUM(new_advertisment_costs.costs) AS total_costs,
                        SUM(new_visits.revenue) - SUM(new_advertisment_costs.costs) AS markt_metric
                    FROM (
                        SELECT
                           device,
                           sourceid,
                           SUM(o.revenue) as revenue, 
                           DATE_TRUNC('day', o.buyts)::date AS date_revenue
                        FROM
                           orders o
                        RIGHT JOIN (SELECT 
                                        uid, 
                                        MAX(sourceid) AS sourceid 
                                    FROM
                                        visits 
                                    GROUP BY 
                                        uid
                                    ) AS table_visits_1 ON table_visits_1.uid = o.uid
                        INNER JOIN (SELECT 
                                        uid, 
                                        MAX(device) AS device 
                                    FROM 
                                        visits 
                                    GROUP BY 
                                        uid
                                    ) AS table_visits_2 ON table_visits_2.uid = o.uid
                        GROUP BY 
                            table_visits_2.device, table_visits_1.sourceid, date_revenue
                                   ) AS new_visits
                    INNER JOIN (SELECT
                                    sourceid, 
                                    SUM(costs) AS costs, 
                                    advertisment_costs.dt::date AS date_costs 
                                FROM 
                                    advertisment_costs
                                GROUP BY 
                                    sourceid, 
                                    date_costs
                               ) AS new_advertisment_costs 
                                 ON new_visits.sourceid = new_advertisment_costs.sourceid 
                                 AND new_visits.date_revenue = new_advertisment_costs.date_costs
 
                    GROUP BY
                        new_visits.device, 
                        new_visits.sourceid,
                        new_visits.date_revenue
                    ORDER BY
                        date, sourceid
                '''

sqlquery(query4)

date   device  sourceid total_revenue total_costs markt_metric
0     2017-06-01  desktop         1       $139.36      $75.20       $64.16
1     2017-06-01  desktop         2         $3.79     $132.56     -$128.77
2     2017-06-01    touch         2        $11.55     $132.56     -$121.01
3     2017-06-01  desktop         3        $71.69     $349.38     -$277.69
4     2017-06-01    touch         3        $10.46     $349.38     -$338.92
...          ...      ...       ...           ...         ...          ...
4656  2018-05-31    touch         5        $61.11     $398.04     -$336.93
4657  2018-05-31  desktop         9        $65.51      $27.37       $38.14
4658  2018-05-31    touch         9        $75.35      $27.37       $47.98
4659  2018-05-31  desktop        10       $309.98      $17.60      $292.38
4660  2018-05-31    touch        10        $37.70      $17.60       $20.10

[4661 rows x 6 columns]

Судя по таблице, второй и третий рекламный источник по мобильной и десктопной версии не окупаются. Затраты превышают выручку. Тоже самое касается рекламного размещения №5 по мобилке.

Рекламные размещения №9 и №10 — окупаются. Видимо, лучше всего отрабатывает рекламный источник №10 по десктопной версии.

Метрика отображает отношение рекламных затрат по данному рекламному источнику к количеству посетителей, перешедших на сайт с данного источника.

In [17]:
query5 ='''   SELECT
                    t.date,
                    t.sourceid,
                    advertisment_costs.costs / t.cnt_users as metrica
            FROM
                    advertisment_costs JOIN
                            (
                                SELECT
                                    DATE_TRUNC('day', startts::timestamp) as date,
                                    count(uid) as cnt_users,
                                    sourceid
                                FROM
                                    visits
                                GROUP BY
                                    sourceid,
                                    date
        ) as t ON advertisment_costs.sourceid = t.sourceid AND 
        DATE_TRUNC('day', advertisment_costs.dt::timestamp) = t.date

ORDER BY 
    t.date, 
    t.sourceid'''

sqlquery(query5)

date  sourceid metrica
0    2017-06-01         1   $1.10
1    2017-06-01         2   $1.55
2    2017-06-01         3   $1.88
3    2017-06-01         4   $0.54
4    2017-06-01         5   $0.60
...         ...       ...     ...
2536 2018-05-31         3   $2.03
2537 2018-05-31         4   $0.96
2538 2018-05-31         5   $1.57
2539 2018-05-31         9   $0.44
2540 2018-05-31        10   $0.13

[2541 rows x 3 columns]

<div style="border:solid blue 2px; padding: 20px"> <h1 style="color:blue; margin-bottom:20px">Комментарий наставника</h1>

Результаты верные, замечательно!

## Еще раз о выводах и общие заключения
<a name="summary"></a>


<div class="col-md-3">
  <ul class="nav flex-column nav-pills">
    <li class="active"><a class="nav-link" id="home-tab" data-toggle="pill" href="#output1" role="tab" aria-controls="output1" aria-selected="true">Посчитайте затраты на каждый рекламный источник.</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#output2" role="tab" aria-controls="output2" aria-selected="false">Посчитайте выручку от заказов, сделанных с настольных ПК (тип устройства desktop)</a></li>
    <li><a class="nav-link" id="profile-tab" data-toggle="pill" href="#output3" role="tab" aria-controls="output3" aria-selected="false">Ввод новой метрики эффективности рекламы - дневные затраты на рекламу на пользователя по источнику</a></li>
    <li><a class="nav-link" id="contact-tab" data-toggle="pill" href="#output" role="tab" aria-controls="output" aria-selected="false">Финальные выводы</a></li>
   
  </ul>
</div>


<div class="tab-content">
<div id="output1" class="tab-pane fade in active">
    <h3 style="margin-left:270px">Посчитайте затраты на каждый рекламный источник.</h3>
    <p style="margin-left:270px"> Всего 10 рекламных источников. Три из них отсутствуют. Возможно, это связано с технической ошибкой или отсутствием переходов с данных рекламных размещений. <br><br> Самые большие суммарные затраты — 141 тыс., связаны с рекламным размещением №3. Далее идёт рекламный источник №4 — 61 тыс. Примерно в равной степени самые минимальные затраты приходятся на рекламные размещения №9 и 10 — 5,5 и 5.8 тыс. соответственно</p>
  </div>

<div id="output2" class="tab-pane fade">
    <h3 style="margin-left:270px">Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег. Сделайте выводы</h3>
    <p style="margin-left:270px">
В целом затраты на рекламный источник № 3 не окупаются с декстопной версии платформы. Разница затрат равна 35,263 тыс. Конверсия выручки от затрат составляет примерно 75%. Несмотря на то, что большая часть вложений приходится на рекламное размещение № 3, на данный момент оно не приносит прибыли.<br><br> Возможно, дополнительная выручка закрывающая часть затрат связанных с рекламным размещением № 3 приходит из мобильной версии платформы. Поэтому, данному рекламному размещению принадлежит большая часть вложений бюджета маркетинга.

</p>
  </div>

  <div id="output3" class="tab-pane fade">
      <h3 style="margin-left:270px">Введём новую метрику эффективности рекламны - дневные затраты на рекламу на пользователя по источнику.</h3>
      <p>
    <p style="margin-left:270px">Судя по таблице, второй и третий рекламный источник по мобильной и десктопной версии не окупаются. Затраты превышают выручку. Тоже самое касается рекламного размещения №5 по мобилке.<br> <br>Рекламные размещения №9 и №10 — окупаются. Видимо, лучше всего отрабатывает рекламный источник №10 по десктопной версии.<br><br>Метрика отображает отношение рекламных затрат по данному рекламному источнику к количеству посетителей, перешедших на сайт с данного источника.
            </p>
  </div>
  
  <div id="output" class="tab-pane fade">
      <h3 style="margin-left:270px">Финальные выводы</h3>
      <p>
        <p style="margin-left:270px">  
          Мы проанализировали стратегию рекламной кампании маркетплейса «Заберу». Она выстраивается на рекламных акциях длительностью <b>1 день</b>. В целом рекламная кампания показала свою неэффективность.<br><br>
Самые большие суммарные затраты — <b>141 тыс.</b>, связаны с рекламным размещением <b>№3</b>. Далее идёт рекламный источник №4 — 61 тыс. Примерно в равной степени самые минимальные затраты приходятся на рекламные размещения <b>№9 и 10</b> — 5,5 и 5,8 тыс. соответственно Основная часть рекламных размещений не окупается. <br><br><b>Видимо, лучше всего отрабатывает рекламный источник №10 по десктопной версии.</b></p>
            </div>